feat(sqlite): Use new SQLite Database

hf-krechan opened the pull request.

hf-krechan committed:
update function to get ahb table from database

hf-krechan committed:
update function to get all format versions

hf-krechan committed:
WIP to update get all known prüfis

hf-krechan committed:
fix query for pruefis

hf-krechan committed:
remove D_ from Dataelements

hf-krechan committed:
update ahb line entity

hf-krechan committed:
add conditions in table view

hf-krechan committed:
remove duplicate SG

hf-krechan committed:
update ahb.db

hf-krechan committed:
put mapping into two functions

hf-krechan committed:
add line_type in orm model

hf-krechan committed:
remove unused import

hf-krechan committed:
fix segment name

hf-krechan committed:
wip column mapping

hf-krechan committed:
add line_type to openapi sepc

hf-krechan committed:
remove unused import

hf-krechan committed:
generate code from openapi spec

hf-krechan committed:
add line_type

hf-krechan committed:
use line type to define border style

hf-krechan committed:
getting closer

hf-krechan committed:
even closer

hf-krechan committed:
remove unused code sections

hf-krechan committed:
use isSegmentGroup

hf-krechan committed:
Change from data_element to data_element: thin line

hf-krechan committed:
refactor broder style matching

hf-krechan committed:
use git lfs

hf-krechan committed:
change filter for lfs to *.db

hf-krechan committed:
add fernet dependency

hf-krechan committed:
use encrypted database

hf-krechan committed:
add encrypted sqlite

hf-krechan committed:
ignore ahb.db

hf-krechan committed:
use lfs true for checkouts in github actions

hf-krechan committed:
ignore 7z files

hf-krechan committed:
add fenret key to dev and prod stacks

hf-krechan committed:
read and set fenret key

hf-krechan committed:
remove unused import

hf-krechan committed:
fix linter warnings by running `ng lint --fix`

hf-krechan committed:
set fernet key env var

hf-kklein commented on src/server/entities/ahb-line.entity.ts:
```diff
@@ -2,19 +2,36 @@
 export class AhbLine {
   @PrimaryColumn({ type: 'varchar', length: 32 })
   id!: string;
 
-  @Column({ type: 'integer' })
-  @Index('ix_ahbline_position_inside_ahb')
-  position_inside_ahb!: number;
+  @Column({ type: 'varchar' })
+  format_version!: string;
```
hier könntest du natürlich auf die stelle verweisen wo das herkommt, also konkret z.b. das ahbmodel bzw. das (cli?) command mit dem diese datenstruktur erstellt wird. jemand, der dann quer einsteigt hat zumindest einen anhaltspunkt.

hf-krechan replied:
Meinst du wo man sieht wie die Datenbank aussieht?

hf-kklein replied:
hier: https://github.com/Hochfrequenz/xml-fundamend-python/blob/main/src/fundamend/sqlmodels/create_ahbtabellen_view.sql oder hier: https://github.com/Hochfrequenz/xml-fundamend-python/blob/9a45e07f953d42c707db4c7af4294790f26e01ce/src/fundamend/sqlmodels/ahbtabellen_view.py#L37-L60

mussdt ja nicht in die tiefe gehen. nur ein hint wo man anfängt zu suchen, wenn dort etwas nicht stimmt.

hf-krechan replied:
7182cc25f0d6eefd38be3e6aa94932a12308d01b

hf-kklein commented on src/server/infrastructure/database.ts:
```diff
@@ -2,13 +2,36 @@ import { DataSource } from 'typeorm';
+  });
+  const decryptedData = token.decode();
+  fs.writeFileSync(decryptedDbPath, decryptedData);
+}
```
copilot says:
> The current decryption logic reads the entire encrypted database into memory (`fs.readFileSync`) and writes the decrypted content to the file using `fs.writeFileSync`. This approach is inefficient for large files, as it requires loading the entire file into memory, which can lead to memory exhaustion for very large files.

To fix this issue and make it work for large files, you can process the file in chunks using streams. Streams allow you to read and write data incrementally, using less memory.

Here’s how you can adjust the decryption logic to work with streams:

```ts
import { createReadStream, createWriteStream } from 'fs';
import { Transform } from 'stream';

if (
  !fs.existsSync(decryptedDbPath) ||
  fs.statSync(encryptedDbPath).mtime > fs.statSync(decryptedDbPath).mtime
) {
  const secret = new fernet.Secret(fernetKey);

  // Create a transform stream to decrypt the chunks
  const decryptStream = new Transform({
    transform(chunk, encoding, callback) {
      const token = new fernet.Token({
        secret: secret,
        token: chunk.toString('base64'),
        ttl: 0,
      });
      const decryptedChunk = token.decode();
      callback(null, decryptedChunk);
    },
  });

  // Create read and write streams
  const readStream = createReadStream(encryptedDbPath);
  const writeStream = createWriteStream(decryptedDbPath);

  // Pipe the read stream through the decrypt stream to the write stream
  readStream.pipe(decryptStream).pipe(writeStream);

  // Handle errors
  readStream.on('error', (err) => {
    console.error('Error reading encrypted file:', err);
  });
  writeStream.on('error', (err) => {
    console.error('Error writing decrypted file:', err);
  });
  writeStream.on('finish', () => {
    console.log('Decryption completed successfully.');
  });
}
```

### Key Changes:
1. **Stream-based Processing**: The `createReadStream` and `createWriteStream` methods allow you to process the file incrementally.
2. **Transform Stream**: A custom `Transform` stream is used to decrypt each chunk of data as it is read.
3. **Error Handling**: Proper error handling is added to detect and log any issues during the read, decrypt, or write process.

### Benefits:
- **Memory-Efficient**: Processes the file in chunks, avoiding memory exhaustion.
- **Scalable**: Suitable for large files since only a small part of the file is in memory at any given time.

This change will allow your decryption logic to handle large files efficiently.


aber ist iwi auch albern, dass ich ai generierte kommentare paste ohne sie überprüft zu haben.

hf-krechan replied:
deprecated comment. We tried the streaming approach but it did not work.
Fernet is not designed for large file encryption: https://cryptography.io/en/latest/fernet/#limitations

hf-kklein commented on src/server/repository/ahb.ts:
```diff
@@ -32,50 +31,62 @@ export default class AHBRepository {
+        ? line.data_element.substring(2)
+        : line.data_element || '',
+      guid: line.id,
+      index: 0, // This will need to be calculated based on sort_path if needed
```
wenn du das brauchst kann ich es dir auch als `row_number()` im `v_ahbesser` view mitgeben, also wenn es dir hier signifikant aufwand spart.

hf-krechan replied:
bin selbst nicht so sicher warum es hier ist.
aber es wäre in Zukunft ggf. praktisch wenn man die Deeplinks hinzufügt.

hf-kklein replied:
Die Deep links sollten besser den path als die rownumber verwenden.

hf-kklein commented on src/server/repository/ahb.ts:
```diff
@@ -32,50 +31,62 @@ export default class AHBRepository {
+      name: isSectionLine ? '' : line.line_name || '',
+      section_name: isSectionLine ? line.line_name || '' : '',
+      segment_code: line.segment_code || '',
+      segment_group_key: line.segmentgroup_key?.startsWith('SGSG')
```
machst du dafür noch ein issue auf im fundamend mit ein, zwei beispielen wo das auftritt? dann kannstes auch hier verlinken.

hf-kklein replied:
fixed in https://github.com/Hochfrequenz/ahbesser/pull/553/commits/a867a83d96126d55894ac78fcd77f87d12229bb4

hf-kklein commented on src/server/repository/formatVersion.ts:
```diff
@@ -57,12 +57,12 @@ export default class FormatVersionRepository extends BlobStorageContainerBacked
-      .where('ahb.edifact_format_version = :formatVersion', { formatVersion })
+      .select('DISTINCT ahb.pruefidentifikator, ahb.description')
+      .where('ahb.format_version = :formatVersion', { formatVersion })
       .orderBy('ahb.pruefidentifikator')
```
ist diese sortierung denn stabil? wenn ja, dann brauchen wir das order-by ws. nicht. wenn nein, dann sollten wir für stabile ergebnisse die description mit den order-by clause aufnehmen.

https://github.com/Hochfrequenz/xml-fundamend-python/blob/9a45e07f953d42c707db4c7af4294790f26e01ce/src/fundamend/sqlmodels/materialize_ahb_view.sql#L499-L500

kannst mal checken, ob das tatsäclich als unique constriant angelegt wird oder nur ein index.
Da gibt's nen kleinen hack: 
https://github.com/Hochfrequenz/xml-fundamend-python/blob/9a45e07f953d42c707db4c7af4294790f26e01ce/src/fundamend/sqlmodels/internals.py#L27-L31

hf-krechan replied:
ich denke ja, es geht hier um die pruefis und die sollten sortiert zurückkommen.
Oder versteh ich die Frage nicht?

hf-kklein replied:
Theoretisch kann es den selben prüfi mit verschiedenen descriptions geben.

hf-kklein reviewed (approved):
ein satz in der readme oder irgendwo, wo man es findet, wäre gut um contributors auf die quelle der DB zu stoßen. wenn du den key z.b. in 1password abgelegt hast, wäre das allein eine bedeutende einstiegshürde weniger. und wenn man den key braucht um überhaupt lokal zu arbeiten, dann wäre es auch cool, wenn das in der readme prominent stünde. können dann ja immer noch den key privat mit contributors teilen, aber es sollte iwo stehen.

hf-krechan committed:
use 7z

hf-krechan committed:
working approach 7z approach

hf-krechan committed:
remove duplicate code

hf-krechan committed:
remove fernet encrypted ahb db

hf-krechan committed:
use new env var name to clarify

hf-krechan committed:
remove fernet dependencies

hf-krechan committed:
update environment variable in pulumi

hf-krechan committed:
update secret for 7z archive

hf-krechan committed:
add environment variable for 7z archive password

hf-krechan committed:
update environment variable in test ci

hf-krechan committed:
add link to secret

hf-krechan committed:
update file name for encrypted database

hf-krechan committed:
use 7zip-min

hf-krechan committed:
feat(7z-decrypt): Use 7z in Dockerfile (#554)

* use 7z in dockerfile

* always replace 7z archive result

hf-krechan committed:
remove 7zip-min dependency

hf-krechan committed:
remove gitattributes for LFS

hf-krechan committed:
remove lfs

hf-krechan committed:
allow 7z files again

hf-krechan committed:
add 7z sqlite database

Commit:
use most recent sqlite db (contains even more segment, doesn't have the 'SGSG' issue)

https://github.com/Hochfrequenz/xml-migs-and-ahbs/releases/tag/v2025.04.12.8z

hf-krechan committed:
chore: Upgrade to Node 23.11 (#555)

* upgrade to v23.11

* remove empty line

* fix  [ERROR] TS-992008: Could not find stylesheet file './fallback-page.component.css'.

[0]
[0] ✘ [ERROR] TS-992008: Could not find stylesheet file './fallback-page.component.css'. [plugin angular-compiler]
[0]
[0]     src/app/shared/components/fallback-page/fallback-page.component.ts:7:14:
[0]       7 │   styleUrls: ['./fallback-page.component.css'],
[0]         ╵               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[0]

* remove fallback-page.component.css

hf-krechan committed:
clean dockerignore

hf-krechan committed:
improve dockerfile

hf-krechan committed:
add section to tell contributors that they need a password

hf-krechan committed:
update docker-compose

hf-krechan committed:
update example env

hf-krechan committed:
add docker environment

hf-krechan committed:
add docker env in bruno

hf-krechan committed:
update companystylesheet

hf-krechan committed:
add hint where the sqlite comes from

hf-krechan committed:
remove deprecated code lines

hf-krechan committed:
remove 7z relict
